File: apps/web/src/lib/training/relink_candidates.ts
/**
 * Pure candidate-selection logic for the workout re-link picker.
 *
 * Re-linking a completed run to a different planned workout must not
 * let one run count toward two workouts' `plan_progress` — so the
 * picker must NOT offer a run that's already linked (auto-matched or
 * manually) to *another* workout. The data layer fetches the owner's
 * runs + the set of run ids already linked anywhere in the owner's
 * plans; this function decides which of those runs are eligible, and
 * in what order, for a given workout.
 *
 * Kept pure (no Supabase, no auth) so it can be unit-tested directly —
 * the data-layer wrapper in `core/data.ts` is exercised by Playwright.
 */

export interface RelinkCandidateRun {
	id: string;
	/** ISO timestamp the run started (`runs.started_at`). */
	started_at: string;
	/** Metres covered (`runs.distance_m`). */
	distance_m: number;
	/** Seconds of moving/elapsed time (`runs.duration_s`). */
	duration_s: number;
}

export interface RelinkFilterInput {
	/** The owner's runs, any order. */
	runs: RelinkCandidateRun[];
	/**
	 * Run ids already linked (via `completed_run_id`) to ANY of the
	 * owner's plan workouts — including the workout being re-linked.
	 */
	linkedRunIds: Iterable<string>;
	/**
	 * The current `completed_run_id` of the workout being re-linked, if
	 * any. It lives in `linkedRunIds` but must stay selectable so the
	 * picker can show the current pick (and re-confirming it is a no-op
	 * rather than a hidden row). Pass null when the workout is unlinked.
	 */
	currentRunId: string | null;
	/** The workout's `scheduled_date` (ISO `YYYY-MM-DD`). */
	scheduledDate: string;
	/**
	 * Half-window in days around `scheduledDate`; a run started within
	 * ±windowDays of the scheduled date is in-window. Defaults to 7.
	 */
	windowDays?: number;
}

const MS_PER_DAY = 86_400_000;
export const DEFAULT_RELINK_WINDOW_DAYS = 7;

/** Calendar-day distance between a run's start and the scheduled date. */
function dayGap(runStartIso: string, scheduledDate: string): number {
	const a = new Date(runStartIso);
	// Compare against the scheduled calendar date at local midnight.
	const [y, m, d] = scheduledDate.split('-').map(Number);
	const scheduled = new Date(y, (m ?? 1) - 1, d ?? 1);
	const runDay = new Date(a.getFullYear(), a.getMonth(), a.getDate());
	return Math.abs(Math.round((runDay.getTime() - scheduled.getTime()) / MS_PER_DAY));
}

/**
 * Eligible re-link candidates for a workout, newest-first.
 *
 * A run is eligible when it is in-window (within ±windowDays of the
 * scheduled date) AND not already linked to a *different* workout. The
 * workout's own current run stays eligible regardless of window so the
 * current pick is always visible.
 */
export function filterRelinkCandidates(
	input: RelinkFilterInput
): RelinkCandidateRun[] {
	const window = input.windowDays ?? DEFAULT_RELINK_WINDOW_DAYS;
	const linked = new Set(input.linkedRunIds);
	// The current run is allowed even though it's in `linked`.
	if (input.currentRunId) linked.delete(input.currentRunId);

	return input.runs
		.filter((r) => {
			// Never offer a run linked to another workout — that would
			// double-count it in plan_progress.
			if (linked.has(r.id)) return false;
			// The current pick is always in; everything else must be
			// inside the date window.
			if (r.id === input.currentRunId) return true;
			return dayGap(r.started_at, input.scheduledDate) <= window;
		})
		.sort(
			(a, b) =>
				new Date(b.started_at).getTime() - new Date(a.started_at).getTime()
		);
}

File: apps/mobile_ios/lib/relink_candidates.dart
/// Pure candidate-selection logic for the workout re-link picker.
///
/// Dart twin of `apps/web/src/lib/training/relink_candidates.ts` — keep
/// in lockstep (algorithm, edge cases, outputs, test counts).
///
/// Re-linking a completed run to a different planned workout must not
/// let one run count toward two workouts' `plan_progress`, so the picker
/// must NOT offer a run already linked (auto-matched or manually) to
/// *another* workout. The service fetches the owner's runs + the set of
/// run ids already linked anywhere in the owner's plans; this function
/// decides which are eligible, and in what order, for a given workout.

class RelinkCandidateRun {
  final String id;

  /// When the run started (`runs.started_at`).
  final DateTime startedAt;

  /// Metres covered (`runs.distance_m`).
  final double distanceM;

  /// Seconds of moving/elapsed time (`runs.duration_s`).
  final int durationS;

  const RelinkCandidateRun({
    required this.id,
    required this.startedAt,
    required this.distanceM,
    required this.durationS,
  });
}

const int kDefaultRelinkWindowDays = 7;

/// Calendar-day distance between a run's start and the scheduled date.
int _dayGap(DateTime runStart, DateTime scheduledDate) {
  final local = runStart.isUtc ? runStart.toLocal() : runStart;
  final runDay = DateTime(local.year, local.month, local.day);
  final scheduled =
      DateTime(scheduledDate.year, scheduledDate.month, scheduledDate.day);
  return (runDay.difference(scheduled).inDays).abs();
}

/// Eligible re-link candidates for a workout, newest-first.
///
/// A run is eligible when it is in-window (within ±[windowDays] of the
/// scheduled date) AND not already linked to a *different* workout. The
/// workout's own current run stays eligible regardless of window so the
/// current pick is always visible.
List<RelinkCandidateRun> filterRelinkCandidates({
  required List<RelinkCandidateRun> runs,
  required Iterable<String> linkedRunIds,
  required String? currentRunId,
  required DateTime scheduledDate,
  int windowDays = kDefaultRelinkWindowDays,
}) {
  final linked = linkedRunIds.toSet();
  // The current run is allowed even though it's in `linked`.
  if (currentRunId != null) linked.remove(currentRunId);

  final out = runs.where((r) {
    // Never offer a run linked to another workout — that would
    // double-count it in plan_progress.
    if (linked.contains(r.id)) return false;
    // The current pick is always in; everything else must be inside
    // the date window.
    if (r.id == currentRunId) return true;
    return _dayGap(r.startedAt, scheduledDate) <= windowDays;
  }).toList();

  out.sort((a, b) => b.startedAt.compareTo(a.startedAt));
  return out;
}
